exotica_core/src/task_map.cpp

```diff
@@ -73,24 +73,34 @@ void TaskMap::Update(Eigen::VectorXdRefConst x, Eigen::VectorXdRef Phi, Eigen::M
     if (jacobian.rows() != TaskSpaceDim() && jacobian.cols() != x.rows())
         ThrowNamed("Jacobian dimension mismatch!");
 
-    // Store original x (needs to be reset later).
-    Eigen::VectorXd x_original(x);
-    Update(x_original, Phi);
-
-    // Backward finite differencing.
+    // Set constants
     constexpr double h = 1e-6;
-    Eigen::VectorXd x_tmp;
-    Eigen::VectorXd phi_original(Phi);
-    for (int i = 0; i < TaskSpaceDim(); ++i)
+    constexpr double h_inverse = 1.0 / h;
+
+    // Compute x/Phi using forward mapping (no jacobain)
+    Update(x, Phi);
+
+    // Backward finite differencing
+    for (int i = 0; i < jacobian.cols(); ++i)
     {
-        x_tmp = x;
-        x_tmp(i) -= h;
-        Update(x_tmp, Phi);
-        jacobian.row(i) = (1 / h) * (phi_original - Phi);
+        // Setup for gradient estimation
+        Eigen::VectorXd x_backward = x;
+        x_backward(i) -= h;
+        Eigen::VectorXd Phi_backward;
```

Comment:
Don't do this in two operations and don't reallocate at every iteration. Define Phi_backward(TaskSpaceDim) outside the for-loop (same for x_backward)